### PR TITLE
fix: transform suggestion data before syncSuggestions to prevent stale merge fields

### PR DESCRIPTION
**Pull request by bisswanger**
## Summary
- `vulnerabilities` and `permissions` audits passed raw scan data straight into `syncSuggestions`, while stored suggestions used a different canonical field shape. The default merge function's shallow spread couldn't reconcile the two shapes, so `dependency_tree` never actually got refreshed on re-scans and stale/duplicate raw fields (`dependencyTree`, `name`, `version`, `acl`, `privileges`, ...) piled up on the suggestion data instead.
- Moved all data transformation into dedicated functions (`toSuggestionData`, `toTooStrongSuggestionData`, `toAdminSuggestionData`) applied to the full input array *before* `syncSuggestions` runs, so both existing and new suggestion data always share one canonical shape and merge cleanly.
- `mapVulnerabilityToSuggestion` / `mapTooStrongSuggestion` / `mapAdminSuggestion` now just wrap already-transformed data 1:1 instead of doing the reshaping themselves.
- `buildKey` (vulnerabilities) is simplified to the single canonical shape it now always receives.

## Test plan
- [x] `npx mocha test/audits/vulnerabilities.test.js test/audits/permissions/permissions.test.js` — 132 passing
- [x] `src/vulnerabilities` and `src/permissions` both at 100% line/branch/function coverage
- [x] `npm run lint` clean
- [x] Added a regression test using real before/after vulnerability scan fixtures (`test/fixtures/vulnerabilities/vulnerability-report-{old,new}-scan.json`) proving a matched suggestion's `dependency_tree` is now correctly refreshed with no leftover duplicate fields, and a dropped component correctly moves to `OUTDATED`

### CODE REVIEW

**Review by MysticatBot — CHANGES_REQUESTED** [review]
Hey @bisswanger,

**Verdict: Request changes** - one latent runtime crash from the null-coalescing regression in buildKey.
**Complexity: HIGH** - large diff (test fixtures inflate line count; actual code change is focused).
_Changes: Moves data transformation into dedicated functions applied before syncSuggestions to fix stale merge fields on re-scans (10 files)._

## Must fix before merge
1. [Important] `buildKey` crashes on `dependency_tree: null` - behavioral regression - `src/vulnerabilities/handler.js:221`

The new destructuring default `dependency_tree: tree = []` only fires on `undefined`, not `null`. If any upstream scanner returns `dependencyTree: null` (a plausible shape - the same scanner uses `ignoredVulnerabilities: null` in its output), the chain `toSuggestionData` -> `buildKey` will throw `TypeError: tree.filter is not a function`.

The old code handled this safely via `item.dependency_tree ?? item.dependencyTree ?? []` (nullish coalescing covers both `null` and `undefined`).

Fix in `toSuggestionData`:
```js
dependency_tree: dependencyTree || [],
```

This mirrors the existing `vulnerabilities || []` pattern two lines above and makes both nullable array fields consistent. Add a test case for `{ name: "x", version: "1", dependencyTree: null }` to cover the edge.

<details>
<summary>Non-blocking (3): minor issues and suggestions</summary>

- nit: trailing space in `'jcr:write '` - `src/permissions/suggestion-data-mapper.js:95` - pre-existing but now lives in the single canonical source; good time to trim it.
- suggestion: `.sort()` mutates the caller's `vulnerabilities` array in-place - `src/vulnerabilities/suggestion-data-mapper.js:44` - use `[...safeVulnerabilities].sort(...)` or `.toSorted()`. Pre-existing behavior, not a regression, but cheap to fix while the file is open.
- suggestion: validate production data has no stored suggestions where `library` is absent but `name` is present before deploying, to confirm the `buildKey` fallback removal is safe.
</details>

---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 2m 0s | <b>Cost:</b> $6.11 | <b>Commit:</b> <code>76846a2da6b0ddbfe9b60569cacb3f8029eec3bc</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

File: src/vulnerabilities/handler.js [hunk 1]
```diff
@@ -178,25 +178,24 @@ export async function extractCodeBucket(context) {
...
- * @param {Object} item - Raw vulnerable component or stored suggestion data.
+ * @param {Object} data - Suggestion data in its canonical shape (see toSuggestionData).
  * @returns {string} - A stable key derived from library name/version and dependency tree.
  */
-export const buildKey = (item) => {
-  const libName = item.library ?? item.name;
-  const libVersion = item.current_version ?? item.version;
-  const tree = item.dependency_tree ?? item.dependencyTree ?? [];
+export const buildKey = ({
+  library, current_version: currentVersion, dependency_tree: tree = [],
```
Comment (MysticatBot) [inline src/vulnerabilities/handler.js]: issue (blocking): `buildKey` destructuring default does not guard against stored `dependency_tree: null`.

The destructuring default `dependency_tree: tree = []` only fires when the value is `undefined`, not `null`. The fix in `toSuggestionData` ensures new data always has an array, but `syncSuggestions` also calls `buildKey` on existing stored suggestions via `getData()`. If any suggestion was persisted before this fix with `dependency_tree: null` (plausible - the same scanner uses `ignoredVulnerabilities: null`), the first audit run after deployment will throw `TypeError: tree.filter is not a function`.

The old code handled this safely via `item.dependency_tree ?? item.dependencyTree ?? []`.

Fix:
```js
export const buildKey = ({
  library, current_version: currentVersion, dependency_tree: tree,
}) => {
  const parts = (tree || [])
    .filter((entry) => entry !== '[root]')
    .map((entry) => entry.replace(/@[^@]*$/, ''));
  return [`${library}@${currentVersion}`, ...parts].join('-');
};
```

This mirrors the `dependencyTree || []` pattern in `toSuggestionData` and makes `buildKey` defensive against any shape it receives from the database.

**Review by MysticatBot — CHANGES_REQUESTED** [review]
Hey @bisswanger,

**Verdict: Request changes** - one remaining null-safety regression in `buildKey` for stored data.
**Complexity: HIGH** - large diff (test fixtures inflate line count; actual code change is focused).
_Changes: Moves data transformation into dedicated functions applied before syncSuggestions to fix stale merge fields on re-scans (10 files)._
Note: CI checks are currently pending - verify before merge.

## Must fix before merge
1. [Important] `buildKey` destructuring default does not guard against stored `dependency_tree: null` - `src/vulnerabilities/handler.js:220` (details inline)

<details>
<summary>Non-blocking (1): minor issues and suggestions</summary>

- suggestion: `.sort()` mutates the caller's `vulnerabilities` array in-place - `src/vulnerabilities/suggestion-data-mapper.js:60` - use `[...safeVulnerabilities].sort(...)` or `.toSorted()`. Pre-existing behavior, not a regression, but cheap to fix while the file is open.
</details>

## Previously flagged, now resolved
- Null dependency tree guard added in `toSuggestionData` (new data path is safe)
- Trailing space in `'jcr:write '` trimmed to `'jcr:write'`

---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 2m 28s | <b>Cost:</b> $5.65 | <b>Commit:</b> <code>bb8a23c1aa2117b6d66872577038a24dd5d2df61</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Review by MysticatBot — APPROVED** [review]
Hey @bisswanger,

**Verdict: Approve** - prior null-safety finding is addressed; no new issues.
**Complexity: HIGH** - large diff (test fixtures inflate line count; actual code change is focused).
_Changes: Adds null-coalescing guard to buildKey for stored dependency_tree: null, with a targeted regression test (10 files)._

<details>
<summary>Non-blocking (1): minor issues and suggestions</summary>

- suggestion: `.sort()` mutates the caller's `vulnerabilities` array in-place - `src/vulnerabilities/suggestion-data-mapper.js:60` - use `[...safeVulnerabilities].sort(...)` or `.toSorted()`. Pre-existing behavior, not a regression, but cheap to fix while the file is open.
</details>

## Previously flagged, now resolved
- Null dependency_tree guard added directly in `buildKey` via `(tree || [])` - handles both null and undefined from stored suggestions
- (Prior review round 1) Null guard in `toSuggestionData` and trailing space in `'jcr:write'` were already resolved

---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 0m 57s | <b>Cost:</b> $3.78 | <b>Commit:</b> <code>876cafe6050b317fc5b555fa2664d10588d50d92</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>
